Remove prim::Constant op

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33265 Clean up isinstance flags
* #32889 Move special ops into interpreter
* **#32804 Remove prim::Constant op**
* #32791 interpreter handling for varargs to remove need for looking at Node

Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

zdevito committed:
Remove prim::Constant op

Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit da7354d6:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=b565d9b35696259f9ab7d5dfd3814c82d40aaa84&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base b565d9b3 since Jan 29
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=da7354d60d3f981bb7079b619df46d0bdececd59).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_backward_compatibility_check_test](https://dr.pytorch.org/build-details.html?build_id=92965795) from Jan 29



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

bhosmer commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -75,39 +75,35 @@ struct ConstantPropagator {
-    } catch (...) {
-      // Catch exceptions. This op may not be run,
-      // so catch the error here & leave the op in the graph
+    if (auto outputs_opt = runNode(n)) {
```
Nice to be rid of try/catch, and I'm assuming it's because Constant was the only op that might throw? But I'm not sure what motivated the optionalizing of the return value.

zdevito replied:
I am localizing the try/catch logic inside of runNode since it is an implementation detail of how we do constant prop by actually run the potentially wrong operator. I moved the getOperator call outside of the try/catch because that was actually a bug: getOperator should always succeed but we were silently letting it fail. It matters in the next PR where I introduce a bunch of things where getOperator would fail.

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]

apaszke commented on torch/csrc/jit/constants.cpp:
```diff
@@ -140,117 +140,63 @@ c10::optional<Value*> tryInsertConstant(
-  return stack.back();
+  const Node* node = v->node();
+  const TypePtr& type = v->type();
+  if (type->isSubtypeOf(TensorType::get())) {
```
Shouldn't we at least check that `node->kind() == prim::Constant` before we start blindly reaching for its attributes?

zdevito committed:
Update on "Remove prim::Constant op"


Constants are interpreter primitives so the op was not actually used.
This cleans up some of the logic around it.

This also fixes constant prop such that failures to look up an op
do not silently stop constant propagation. Instead, only errors
inside the op implementation itself will do this.

Differential Revision: [D19673156](https://our.internmc.facebook.com/intern/diff/D19673156)

[ghstack-poisoned]
